Deprecation documentation (#4752)
project-koku/koku: docs/team_workflows/deprecating_an_endpoint.md

# Deprecation of Endpoint Workflow

## Introduction

This document outlines the workflow to deprecate an endpoint within koku. Deprecating an endpoint involves several steps to ensure a smooth transition for users and to maintain the integrity of our services.

## Workflow

### Step 1: Impact Analysis

- **Objective:** Understand the impact of deprecating the endpoint on users, applications, and systems.
- **Actions:**
    - Identify users or services currently using the endpoint.
    - Analyze the extent of usage and dependencies on the endpoint.
    - Determine if there are alternative endpoints or methods to replace the deprecated one.
- **Example:**

>     - When we deprecated the /settings endpoint we made a kibaba dashboard to show us which users were still hitting the endpoint after the old UI was removed. That informed us that users were utilizes those urls in there workflows.

Reviewer: ```suggestion
    - When we deprecated the /settings endpoint we made a kibana dashboard to show us which users were still hitting the endpoint after the old UI was removed. That informed us that users were utilizes those urls in there workflows.
```

### Step 2: Create a Deprecation Timeline

- **Objective:** Define a timeline for the deprecation process.
- **Actions:**
    - Create a timeline that includes the announcement date, start of the deprecation period, and the official deprecation date.
    - Identify any milestones or checkpoints during the deprecation period.

### Step 3: Deprecation Implementation

> - **Objective:** Prepare the codedate for deprecation

Reviewer: ```suggestion
- **Objective:** Prepare the code date for deprecation
```

- **Actions:**

> 1. Renamed the api folder to have the `deprecated_` to clearly identify no changes should be made in this directory moving forward.

Reviewer: ```suggestion
1. Rename the api folder adding `deprecated_` to clearly identify no changes should be made in this directory moving forward.
```

2. Check for unleash flags within the folder and remove any unnecessary flags

> 3. Moved any common functions that were no longer necessary into the deprecated_ folder. [Example](https://github.com/project-koku/koku/pull/4670/files#diff-9a4ece704604756b549b41d1e4b72e154ba290f2ba90cb032d270e5a9e190418L390-L440)

Reviewer: ```suggestion
3. Move any common functions that are no longer necessary into the deprecated_ folder. [Example](https://github.com/project-koku/koku/pull/4670/files#diff-9a4ece704604756b549b41d1e4b72e154ba290f2ba90cb032d270e5a9e190418L390-L440)
```

> 4. Create `deprecated_` versions of any common functions till being used in the `deprecated_` folder, this prevents the need for backwards compatability moving forward. [Example](https://github.com/project-koku/koku/pull/4670/files#diff-3f27604615bb126d6ad77ab747562f4fb028d4975a8cc2ab28a3967ea8f82a88R242-R266)

Reviewer: ```suggestion
4. Create `deprecated_` versions of any common function still being used in the `deprecated_` folder, this prevents the need for backwards compatibility moving forward. [Example](https://github.com/project-koku/koku/pull/4670/files#diff-3f27604615bb126d6ad77ab747562f4fb028d4975a8cc2ab28a3967ea8f82a88R242-R266)
```

5. Add the deprecation wrapper

>     - First update the view to include the `deprecation_datetime` & `sunset_datetime` datetime. [Example](https://github.com/project-koku/koku/blob/9f46c8c3db6d856558ffdc9cf823ab6116590c67/koku/api/deprecated_settings/view.py#L20-L28)

Reviewer: ```suggestion
    - First update the view to include the `deprecation_datetime` & `sunset_datetime`. [Example](https://github.com/project-koku/koku/blob/9f46c8c3db6d856558ffdc9cf823ab6116590c67/koku/api/deprecated_settings/view.py#L20-L28)
```
